### PR TITLE
Initial Quarkus Setup

### PR DESCRIPTION
**Pull request by gr812b**
Begun with initial quarkus setup, which includes uploading and downloading files, some application properties and more

### CODE REVIEW

**Review by camdnnn — CHANGES_REQUESTED** [review]
Old API folder should be deleted

File: backend/src/main/java/com/mcmasterbaja/services/FileMetadataService.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+   */
+  LocalDateTime getZeroTime(Path folderPath);
+
+  /**
+   * Gets the desired folder for a file.
+   *
+   * @param pathString The Path of the file to analyze.
+   * @return The desired type folder.
+   */
+  String getTypeFolder(String pathString);
```
Comment (gr812b) [inline backend/src/main/java/com/mcmasterbaja/services/FileMetadataService.java:72]: Despite passing a String in its one use case, every other method deals almost solely in Paths for paths, which is a much more robust type for this anyways, so I think this should be Path

File: backend/src/test/java/com/mcmasterbaja/GreetingResourceIT.java
Comment (gr812b) [inline backend/src/test/java/com/mcmasterbaja/GreetingResourceIT.java:1]: Tests should be updated or removed

**Review by gr812b — COMMENTED** [review]
- Worth looking into Quarkus' `@RestQuery` instead of jakarta's `@PathParam`
- Exceptions feel a little scuffed around the edges. A sit down to define inheritance, local variables, how to fix errors, feedback, etc
- Are the CI actions updated?
- Does XYColour work?

File: backend/src/main/java/com/mcmasterbaja/binary_csv/BinaryToCSV.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import java.nio.file.Paths;
+
+public class BinaryToCSV {
+
+  public static native void toCSV(String filename, String outputDir, boolean folder);
+
+  public static native void bytesToCSV(
+      byte[] bytes, String outputDir, String fileName, boolean folder);
+
+  private static final String relativePath = "/src/main/java/com/mcmasterbaja/binary_csv/";
```
Comment (JAWalmsley) [inline backend/src/main/java/com/mcmasterbaja/binary_csv/BinaryToCSV.java:12]: forward slashes are ruining my life on windows, use Path.of() to parse it properly for each OS

**Review by JAWalmsley — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/mcmasterbaja/binary_csv/BinaryToCSV.java (hunk 2 above)
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/binary_csv/BinaryToCSV.java:12]: These forward slashes were the same on the old one, are you sure this is the problem?

Comment (JAWalmsley) [reply backend/src/main/java/com/mcmasterbaja/binary_csv/BinaryToCSV.java:12]: nevermind, was unrelated apparently

**Review by JAWalmsley — APPROVED** [review]

File: backend/src/main/java/com/mcmasterbaja/FileFetchResource.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    List<String> fileNames =
+        storageService.loadAll().map(Path::toString).collect(Collectors.toList());
+
+    return fileNames;
+  }
+
+  // TODO: What exception is thrown when it can't find the file?
+  @GET
+  @jakarta.ws.rs.Path("/{filekey}")
+  public RestResponse<File> getFile(@PathParam("filekey") String filekey) {
```
[the remark this answers is not in the document]
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/FileFetchResource.java]: It did not.

File: backend/src/main/java/com/mcmasterbaja/FileFetchResource.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    logger.info("Getting timespan for folder: " + folderkey);
+
+    Path folderPath = Paths.get(folderkey);
+    List<FileTimespan> timespans = new ArrayList<>();
+    Stream<Path> paths = storageService.loadAll(folderPath);
+
+    switch (folderkey) {
+      case "csv":
+        // Holds the parent folder and the zero time
+        Object[] container = {null, null};
```
[the remark this answers is not in the document]
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/FileFetchResource.java:122]: @camdnnn

File: backend/src/main/java/com/mcmasterbaja/analyzer/AnalyzerFactory.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+public class AnalyzerFactory {
+
+  public static Analyzer createAnalyzer(AnalyzerParams params) {
+
+    String[] inputFiles = params.getInputFiles();
+    String[] inputColumns = params.getInputColumns();
+    String[] outputFiles = params.getOutputFiles();
+    Object[] options = params.getOptions();
+
+    switch (params.getType()) {
```
[the remark this answers is not in the document]
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/analyzer/AnalyzerFactory.java:14]: Nahh, that sounds silly

File: backend/src/main/java/com/mcmasterbaja/live/Serial.java
[the remark this answers is not in the document]
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/live/Serial.java:1]: Created https://github.com/McMaster-Baja-Racing/Better-Data-Viewer/issues/120 to address this.

File: backend/src/main/java/com/mcmasterbaja/model/ErrorResponse.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+@Getter
+@EqualsAndHashCode
+@ToString
+public class ErrorResponse {
+
+  @JsonInclude(JsonInclude.Include.NON_NULL)
```
[the remark this answers is not in the document]
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/model/ErrorResponse.java]: It removes non null fields from the response goofy

File: backend/src/main/java/com/mcmasterbaja/services/FileMetadataService.java (hunk 1 above)
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/services/FileMetadataService.java]: Fixed this

File: start.ps1
[the remark this answers is not in the document]
Comment (gr812b) [reply start.ps1:1]: It works now

File: backend/src/main/java/com/mcmasterbaja/FileAnalyzeResource.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+    Path targetPath = storageService.load(Paths.get(filekey));
+    Double[] minMax = fileMetadataService.getMinMax(targetPath, column);
+
+    return minMax;
+  }
+
+  @PATCH
+  @jakarta.ws.rs.Path("togglelive")
+  public String toggleLive() {
```
[the remark this answers is not in the document]
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/FileAnalyzeResource.java:82]: It works now

**Comment by gr812b** [comment]
XYColour and CI Actions working in full, all other features seem to still work as well

File: backend/src/main/java/com/mcmasterbaja/FileFetchResource.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (camdnnn) [reply backend/src/main/java/com/mcmasterbaja/FileFetchResource.java]: I replaced the list with a mapping of Parent to ZeroTime which ended up working quite nicely. If we could store this map somewhere globally we wouldn't need to recompute zeroTimes

**Review by camdnnn — APPROVED** [review]

**Pull request merged**